File: src/main/cpp/g0001_0100/s0094_binary_tree_inorder_traversal/Solution.cpp
// #Easy #Top_100_Liked_Questions #Top_Interview_Questions #Depth_First_Search #Tree #Binary_Tree
// #Stack #Data_Structure_I_Day_10_Tree #Udemy_Tree_Stack_Queue #Big_O_Time_O(n)_Space_O(n)
// #2024_05_26_Time_0_ms_(100.00%)_Space_9.8_MB_(40.16%)

/*
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> a;
    void inord(TreeNode* root)
    {
        if(!root) return;
        if(root->left) inord(root->left);
        a.push_back(root->val);
        if(root->right) inord(root->right);
    }
    vector<int> inorderTraversal(TreeNode* root) {
        inord(root);
        return a;
    }
};

File: src/main/dart/g0001_0100/s0094_binary_tree_inorder_traversal/Solution.dart
// #Easy #Top_100_Liked_Questions #Top_Interview_Questions #Depth_First_Search #Tree #Binary_Tree
// #Stack #Data_Structure_I_Day_10_Tree #Udemy_Tree_Stack_Queue #Big_O_Time_O(n)_Space_O(n)
// #2024_10_06_Time_287_ms_(95.71%)_Space_151.5_MB_(11.43%)

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Stack<T> {
  final List<T> _items = [];

  void push(T item) {
    _items.add(item);
  }

  T pop() {
    return _items.removeLast();
  }

  T peek() {
    return _items.last;
  }

  bool get isEmpty {
    return _items.isEmpty;
  }

  int get size {
    return _items.length;
  }
}

class Solution {

  List<int> inorderTraversal(TreeNode? root) {
    final result = List<int>.empty(growable: true);
    final stack = Stack<TreeNode>();

    while (root != null || !stack.isEmpty) {
      while (root != null) {
        stack.push(root);
        root = root.left;
      }
      root = stack.pop();

      result.add(root.val);

      root = root.right;
    }

    return result;
  }
}
